### PR TITLE
update FSDP and add XLA-FSDP documentation

### PR DESCRIPTION
**Pull request by pacman100**
# What does this PR do?

1. update FSDP and add XLA-FSDP documentation

### CODE REVIEW

File: docs/source/en/main_classes/trainer.mdx [hunk 1]
```diff
@@ -574,22 +574,59 @@ add `--fsdp "full_shard offload"` or `--fsdp "shard_grad_op offload"` to the com
...
-  - For transformer based auto wrap policy, please add `--fsdp_transformer_layer_cls_to_wrap <value>` to command line arguments.
-  This specifies the transformer layer class name (case-sensitive) to wrap ,e.g, `BertLayer`, `GPTJBlock`, `T5Block` ....
-  This is important because submodules that share weights (e.g., embedding layer) should not end up in different FSDP wrapped units. 
-  Using this policy, wrapping happens for each block containing Multi-Head Attention followed by couple of MLP layers. 
-  Remaining layers including the shared embeddings are conveniently wrapped in same outermost FSDP unit.
-  Therefore, use this for transformer based models.
-  - For size based auto wrap policy, please add `--fsdp_min_num_params <number>` to command line arguments.
-  It specifies FSDP's minimum number of parameters for auto wrapping.
+- Remaining FSDP config is passed via `--fsdp_config <path_to_fsdp_config.json>`. It is either a location of
+FSDP json config file (e.g., `fsdp_config.json`) or an already loaded json file as `dict`.
```
Comment (sgugger) [inline docs/source/en/main_classes/trainer.mdx:578]: This needs to be on the line below to be in the same list item, or to be indented by 2 spaces.

File: docs/source/en/main_classes/trainer.mdx [hunk 2]
```diff
@@ -574,22 +574,59 @@ add `--fsdp "full_shard offload"` or `--fsdp "shard_grad_op offload"` to the com
...
-  It specifies FSDP's minimum number of parameters for auto wrapping.
+- Remaining FSDP config is passed via `--fsdp_config <path_to_fsdp_config.json>`. It is either a location of
+FSDP json config file (e.g., `fsdp_config.json`) or an already loaded json file as `dict`. 
+  - If auto wrapping is enabled, you can either use transformer based auto wrap policy or size based auto wrap policy.
+    - For transformer based auto wrap policy, please specify `fsdp_transformer_layer_cls_to_wrap` in the config file. 
+    This specifies the list of transformer layer class name (case-sensitive) to wrap ,e.g, [`BertLayer`], [`GPTJBlock`], [`T5Block`] ....
+    This is important because submodules that share weights (e.g., embedding layer) should not end up in different FSDP wrapped units.
+    Using this policy, wrapping happens for each block containing Multi-Head Attention followed by couple of MLP layers. 
+    Remaining layers including the shared embeddings are conveniently wrapped in same outermost FSDP unit.
+    Therefore, use this for transformer based models.
```
Comment (sgugger) [inline docs/source/en/main_classes/trainer.mdx:585]: All intdentation here is wrong.

File: docs/source/en/main_classes/trainer.mdx [hunk 3]
```diff
@@ -574,22 +574,59 @@ add `--fsdp "full_shard offload"` or `--fsdp "shard_grad_op offload"` to the com
...
+**Usage**:
+
+Pass `--fsdp True` along with following changes to be made in `--fsdp_config <path_to_fsdp_config.json>`:
+- `xla` should be set to `True` to enable PyTorch/XLA FSDP.
+- `xla_fsdp_settings` The value is a dictionary which stores the XLA FSDP wrapping parameters.
+  For a complete list of options, please see [here](
+  https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py).
+- `xla_fsdp_grad_ckpt`. When `True`, uses gradient checkpointing over each nested XLA FSDP wrapped layer. This setting can only be
+used when the xla flag is set to true, and an auto wrapping policy is specified through
+`fsdp_min_num_params` or `fsdp_transformer_layer_cls_to_wrap`.
```
Comment (sgugger) [inline docs/source/en/main_classes/trainer.mdx:619]: Indentation is wrong.

File: docs/source/en/main_classes/trainer.mdx [hunk 4]
```diff
@@ -574,22 +574,59 @@ add `--fsdp "full_shard offload"` or `--fsdp "shard_grad_op offload"` to the com
...
+    If `"True"`, FSDP explicitly synchronizes the CPU thread to prevent too many in-flight all-gathers.
 
 **Few caveats to be aware of**
-- Mixed precision is currently not supported with FSDP as we wait for PyTorch to fix support for it.
-More details in this [issues](https://github.com/pytorch/pytorch/issues/75676).
-- FSDP currently doesn't support multiple parameter groups. 
-More details mentioned in this [issue](https://github.com/pytorch/pytorch/issues/76501)
-(`The original model parameters' .grads are not set, meaning that they cannot be optimized separately (which is why we cannot support multiple parameter groups)`).
+- This feature is incompatible with `--predict_with_generate` in the _run_translation.py_ script.
+This feature is incompatible with `generate()` method of `PreTrainedModel` class.
```
Comment (sgugger) [inline docs/source/en/main_classes/trainer.mdx:598]: This should be the first sentence: it is incompatible with `generate`, thus is incompatible with `--predict_with_generate`  in all seq2seq scripts (translation/summarization etc.)

**Review by sgugger — COMMENTED** [review]
Thanks for your work on this, I left a few comment.

File: docs/source/en/main_classes/trainer.mdx (hunk 3 above)
Comment (pacman100) [reply docs/source/en/main_classes/trainer.mdx]: Done.

File: docs/source/en/main_classes/trainer.mdx (hunk 2 above)
Comment (pacman100) [reply docs/source/en/main_classes/trainer.mdx]: Fixed.

File: docs/source/en/main_classes/trainer.mdx (hunk 1 above)
Comment (pacman100) [reply docs/source/en/main_classes/trainer.mdx]: Fixed

**Review by sgugger — APPROVED** [review]

**Comment by JackCaoG** [comment]
Thanks @pacman100 and @sgugger ! This is exciting!

**Pull request merged**